pkg/params/params.go

```diff
@@ -0,0 +1,231 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+Package params provides a generic way to describe parameters used by gadgets, operators
+and runtimes including validation. They can easily be serialized and handed over to different
+frameworks like cobra for use in CLI or a webinterface using JSON.
+*/
+package params
+
+import (
+	"fmt"
+	"strings"
+
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+type (
+	Params           []*Param
+	ParamMap         map[string]string
+	ParamsCollection map[string]Params
+)
+
+type Param struct {
+	// Key is the name under which this param is registered; this will also be the key when
+	// getting a key/value map
+	Key string `json:"key"`
+
+	// Alias is a shortcut for this parameter, usually a single character used for command line
+	// interfaces
+	Alias string `json:"alias"`
+
+	// Title is an optional (pretty) alternative to key and used in user interfaces
+	Title string `json:"title"`
+
+	// DefaultValue is the value that will be used if no other value has been assigned
+	DefaultValue string `json:"defaultValue"`
+
+	// Description holds an optional explanation for this parameter; shown in user interfaces
+	Description string `json:"description"`
+
+	// IsMandatory will be considered when validating; if the param has no value assigned and
+	// also no DefaultValue is set, validation will fail
+	IsMandatory bool `json:"isMandatory"`
+
+	// Tags can be used to skip parameters not needed for a specific environment
+	Tags []string `json:"tags"`
+
+	// Validator is an optional function that will be called upon validation; may or may
+	// not be called in user interfaces. Setting TypeHint is preferred, but can also be used
+	// in combination with the Validator.
+	Validator ParamValidator `json:"-"`
+
+	// TypeHint is the preferred way to set the type of this parameter as it will invoke a
+	// matching validator automatically; if unset, a value of "string" is assumed
+	TypeHint TypeHint `json:"type"`
+
+	// PossibleValues holds all possible values for this parameter and will be considered
+	// when validating
+	PossibleValues []string `json:"possibleValues"`
+
+	value    string
+	assigned bool
+}
+
+// GetTitle returns a human friendly title of the field; is no Title has been specified,
+// the Key will be used with the first letter upper-cased
+func (p *Param) GetTitle() string {
+	if p.Title != "" {
+		return p.Title
+	}
+	return cases.Title(language.English).String(p.Key)
+}
+
+// PreValidate validates a string against the given parameter
+func (p *Param) PreValidate(value string) error {
```

Comment:
"Pre", because it's happening on a value that has not yet been set - whereas Validate() would be performed on the set value. But I agree that this is confusing - a Validate() like I described shouldn't be necessary, if invalid values couldn't be set in the first place.

Reply:
> "Pre", because it's happening on a value that has not yet been set
>
It doesn't seems that obvious to me hahahaha

Reply:
Removed the original Validate and changed it to what PreValidate was.